src/AI/SpenderAI.java

```diff
@@ -0,0 +1,350 @@
+import java.util.Map.Entry;
+import java.util.Queue;
+import CommandingOfficers.Commander;
+import CommandingOfficers.CommanderAbility;
+import Engine.GameAction;
+import Engine.GameAction.ActionType;
+import Engine.GameActionSet;
+import Engine.Path;
+import Engine.Utils;
+import Engine.XYCoord;
+import Terrain.GameMap;
+import Terrain.Location;
+import Terrain.TerrainType;
+import Units.Unit;
+import Units.UnitModel;
+
+/**
+ *  This AI's intent is to just spend all of its resources, action economy included
+ */
+public class SpenderAI implements AIController
+{
+  Queue<GameAction> actions = new ArrayDeque<GameAction>();
+  Queue<Unit> unitQueue = new ArrayDeque<Unit>();
+  boolean stateChange;
+
+  private Commander myCo = null;
+
+  private ArrayList<XYCoord> unownedProperties;
+  private ArrayList<XYCoord> capturingProperties;
+
+  private StringBuffer logger = new StringBuffer();
+  private int turnNum = 0;
+
+  public SpenderAI(Commander co)
+  {
+    myCo = co;
+  }
+
+  @Override
+  public void initTurn(GameMap gameMap)
+  {
+    turnNum++;
+    log(String.format("[======== SpAI initializing turn %s for %s =========]", turnNum, myCo));
+
+    // Make sure we don't have any hang-ons from last time.
+    actions.clear();
+
+    // Create a list of every property we don't own, but want to.
+    unownedProperties = new ArrayList<XYCoord>();
+    for( int x = 0; x < gameMap.mapWidth; ++x )
+    {
+      for( int y = 0; y < gameMap.mapHeight; ++y )
+      {
+        XYCoord loc = new XYCoord(x, y);
+        if( gameMap.getLocation(loc).isCaptureable() && myCo.isEnemy(gameMap.getLocation(loc).getOwner()) )
+        {
+          unownedProperties.add(loc);
+        }
+      }
+    }
+    capturingProperties = new ArrayList<XYCoord>();
+    for( Unit unit : myCo.units )
+    {
+      if( unit.getCaptureProgress() > 0 )
+      {
+        capturingProperties.add(unit.getCaptureTargetCoords());
+      }
+    }
+
+    // If the CO has enough AP, preload the CommanderAbilityAction.
+    ArrayList<CommanderAbility> abilities = myCo.getReadyAbilities();
+    if( abilities.size() > 0 )
+    {
+      actions.offer(new GameAction.AbilityAction(abilities.get(0)));
+    }
+  }
+
+  @Override
+  public void endTurn()
+  {
+    log(String.format("[======== SpAI ending turn %s for %s =========]", turnNum, myCo));
+    System.out.println(logger.toString());
+    logger = new StringBuffer();
+  }
+
+  private void log(String message)
+  {
+    logger.append(message).append('\n');
+  }
+
+  @Override
+  public GameAction getNextAction(GameMap gameMap)
+  {
+    GameAction nextAction = null;
+    do
+    {
+      // If we have more actions ready, don't bother calculating stuff.
+      if( !actions.isEmpty() )
+      {
+        return actions.poll();
+      }
+      else if( unitQueue.isEmpty() )
+      {
+        stateChange = false; // There's been no gamestate change since we last iterated through all the units, since we're about to do just that
```

Comment:
Am I wrong in thinking that `stateChange` will pretty much always have the opposite value of `actions.isEmpty()`? It looks like you pretty much always check and set them together as well.

Reply:
You are wrong.
stateChange requires the unitQueue be emptied and refilled, and the unitQueue is retained between calls of getNextAction().
When stateChange is set to false and *stays* false after re-evaluating unitQueue, we start waiting units and building things.

Reply:
Hmm, I tried it for kicks, and it definitely needs to be there. Good enough.